### PR TITLE
DAOS-6222 dfs: add container create flags for POSIX consistency 

### PR DESCRIPTION
**Pull request by mchaarawi**
- Add relaxed & balanced flags for dfs_cont_create.
- default of those modes will be relaxed in DAOS 2.0
- update the daos tool to be able to set those flags on cont create
- add options to dfs_mount flags to set the mount flags
  - in this case, a container created with balanced can only be
    accessed with balanced mode.
  - a container created with relaxed mode can be accessed with
    either mode.
- add some unit tests to verify behavior.

Signed-off-by: Mohamad Chaarawi <mohamad.chaarawi@intel.com>

### CODE REVIEW

File: src/client/dfs/dfs.c [hunk 1]
```diff
@@ -773,17 +779,31 @@ check_access(dfs_t *dfs, uid_t uid, gid_t gid, mode_t mode, int mask)
...
-	if (flags & O_CREAT) {
-		bool oexcl = flags & O_EXCL;
+	/*
+	 * we only need a DTX in the case of O_CREAT without O_EXCL since we
+	 * don't use a conditional insert.
+	 */
+	if (ocreat && !oexcl && dfs->use_dtx) {
+		rc = daos_tx_open(dfs->coh, &th, 0, NULL);
+		if (rc) {
+			D_ERROR("daos_tx_open() failed (%d)\n", rc);
```
Comment (daltonbohning) [inline src/client/dfs/dfs.c:800]: Formatted DER?
```suggestion
			D_ERROR("daos_tx_open() failed "DF_RC"\n", DP_RC(rc));
```

**Review by daltonbohning — COMMENTED** [review]
Overall, this looks okay to me. But I'd like to look at it again tomorrow with fresh eyes to make sure I understand the transaction logic

File: src/client/dfs/dfs.c [hunk 2]
```diff
@@ -831,49 +851,51 @@ open_file(dfs_t *dfs, daos_handle_t th, dfs_obj_t *parent, int flags,
...
 
 		/** Create and insert entry in parent dir object. */
 		entry->mode = file->mode;
 		entry->atime = entry->mtime = entry->ctime = time(NULL);
 		entry->chunk_size = chunk_size;
 
-		rc = insert_entry(parent->oh, th, file->name, len, entry);
+		rc = insert_entry(parent->oh, th, file->name, len,
+				  (!dfs->use_dtx || oexcl) ?
+				  DAOS_COND_DKEY_INSERT : 0, entry);
```
Comment (daltonbohning) [inline src/client/dfs/dfs.c:864]: What happens if we pass flags=0 (use_dtx) and the entry already exists? In this case, rc=0, so we commit and return. But will entry->mode, entry->chunk_size, etc. be in sync? I.e. the entry already exists, but with different values.

Comment (mchaarawi) [reply src/client/dfs/dfs.c:864]: if the entry exists, we won't be here. would be caught by the fetch.
if it was inserted by someone else after the fetch was called above, then the commit will fail with DER_RESTART, we will restart and the fetch will return exist, and we open the existing file. that is all of course if we do not have o_excl

**Review by daltonbohning — DISMISSED** [review]

**Comment by johannlombardi** [comment]
Is that rebased on mjmac's daos utility change?

File: src/client/dfs/dfs.c [hunk 3]
```diff
@@ -884,6 +906,7 @@ open_file(dfs_t *dfs, daos_handle_t th, dfs_obj_t *parent, int flags,
 	if (rc != 0) {
 		D_ERROR("daos_array_open_with_attr() failed, "DF_RC"\n",
 			DP_RC(rc));
+		D_GOTO(out, rc = daos_der2errno(rc));
 		return daos_der2errno(rc);
```
Comment (ashleypittman) [inline src/client/dfs/dfs.c:910]: Unreachable code?  I'm surprised this got past the compiler to be honest unless I'm misreading it.

File: src/utils/daos.c [hunk 4]
```diff
@@ -881,6 +907,11 @@ common_op_parse_hdlr(int argc, char *argv[], struct cmd_args_s *ap)
 
 	cmd_args_print(ap);
 
+	if (posix_mode_set && ap->type != DAOS_PROP_CO_LAYOUT_POSIX) {
+		fprintf(stderr, "--mode is valid only for a POSIX container\n");
+		D_GOTO(out_free, rc = RC_NO_HELP);
+	}
+
```
Comment (ashleypittman) [inline src/utils/daos.c:914]: This might need updating for the go implementation.

**Review by ashleypittman — DISMISSED** [review]

**Comment by mchaarawi** [comment]
> Is that rebased on mjmac's daos utility change?

yes

File: src/client/dfs/dfs.c (hunk 3 above)
Comment (mchaarawi) [reply src/client/dfs/dfs.c:910]: yea that's a merge conflict that i didn't resolve properly. sigh

File: src/utils/daos.c (hunk 4 above)
Comment (mchaarawi) [reply src/utils/daos.c:914]: im not sure what you mean. can you suggest the change?

Comment (ashleypittman) [reply src/utils/daos.c:914]: I was thinking that this file isn't used any more other than for the daos_old command, and the change here will need to be ported to the new go code.

Comment (mchaarawi) [reply src/utils/daos.c:914]: OK i didn't realize we replaced the entire daos.c main function. 
i guess i will also have missed the mode option in the go frontend..

Comment (mchaarawi) [reply src/utils/daos.c:914]: it seems @mjmac had already copied this from my PR and added it to the go container frontend.
so i think we are good as it's only under type posix:
https://github.com/daos-stack/daos/commit/c66eef2775df7e322213906f2ae7ab6f29c3d16d#diff-c8c509b1b2b0c108d3c02f9e16de02a07c81ad5a920d5c72617198a5bc7b548dR185

Comment (ashleypittman) [reply src/utils/daos.c:914]: I think so, I've got the same issue with https://github.com/daos-stack/daos/pull/5969

Given this is disabled by default it shouldn't block this PR however but you do want to test this.

Comment (mchaarawi) [reply src/utils/daos.c:914]: im going to add some NLT tests for this in a different PR

**Review by daltonbohning — APPROVED** [review]

File: src/utils/daos.c (hunk 4 above)
Comment (jolivier23) [reply src/utils/daos.c:914]: Is this file dead code?

**Review by jolivier23 — APPROVED** [review]

File: src/utils/daos.c (hunk 4 above)
Comment (daltonbohning) [reply src/utils/daos.c:914]: It's `daos_old` now:
https://github.com/daos-stack/daos/blob/c66eef2775df7e322213906f2ae7ab6f29c3d16d/src/utils/SConscript#L45-L49
Presumably to be removed at some later date?

**Pull request merged**